lib/EventListener/FileEventsListener.php

```diff
@@ -0,0 +1,123 @@
+		private TagService $tagService,
+		private IAppConfig $appConfig,
+		private MailService $mailService
+	) {
+	}
+
+	public static function register(IRegistrationContext $context): void {
+		$context->registerEventListener(NodeWrittenEvent::class, self::class);
+	}
+
+	public function handle(Event $event): void {
+		if ($event instanceof NodeWrittenEvent) {
+			$node = $event->getNode();
+			if ($node->getType() !== \OCP\Files\FileInfo::TYPE_FILE) {
+				return;
+			}
+			try {
+				$verdict = $this->verdictService->scanFileById($node->getId());
+			} catch (\Exception $e) {
+				$unscannedTagIsDisabled = $this->appConfig->getValueBool(Application::APP_ID, 'disableUnscannedTag');
+				if (!$unscannedTagIsDisabled) {
+					$this->tagService->setTag($node->getId(), TagService::UNSCANNED, silent: true);
+				}
+				$this->logger->error("Failed to scan uploaded file '{$node->getName()}' with ID '{$node->getId()}': {$e->getMessage()}");
+				return;
+			}
+
+			if ($verdict->Verdict->value == TagService::MALICIOUS) {
+				$this->sendErrorResponse(new VirusFoundException($verdict, $node->getName(), $node->getId()));
+				$this->fileService->deleteFile($node->getId());
+				if ($this->appConfig->getValueBool(Application::APP_ID, 'sendMailOnVirusUpload')) {
+					$this->mailService->notifyMaliciousUpload($verdict, $node->getPath(), $this->userSession->getUser()->getUID(), $node->getSize());
+				}
+				exit;
+			}
+		}
+	}
+
+	public function generateBody(Exception $ex): mixed {
+		if ($this->acceptHtml()) {
+			$templateName = 'exception';
+			$renderAs = 'guest';
+			$templateName = '425';
+		} else {
+			$templateName = 'xml_exception';
+			$renderAs = null;
+			$this->server->httpResponse->setHeader('Content-Type', 'application/xml; charset=utf-8');
+		}
+
+		$debug = $this->config->getSystemValueBool('debug', false);
+
+		$content = new OC_Template('gdatavaas', $templateName, $renderAs);
+		$content->assign('title', 'Error');
+		$content->assign('message', $ex->getMessage());
+		$content->assign('remoteAddr', $this->request->getRemoteAddress());
+		$content->assign('requestID', $this->request->getId());
+		$content->assign('debugMode', $debug);
+		$content->assign('errorClass', get_class($ex));
+		$content->assign('errorMsg', $ex->getMessage());
+		$content->assign('errorCode', $ex->getCode());
+		$content->assign('file', $ex->getFile());
+		$content->assign('line', $ex->getLine());
+		$content->assign('exception', $ex);
+		$contentString = $content->fetchPage();
+		return $contentString;
+	}
+
+	private function acceptHtml(): bool {
+		foreach (explode(',', $this->request->getHeader('Accept')) as $part) {
+			$subparts = explode(';', $part);
+			if (str_ends_with($subparts[0], '/html')) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	private function sendErrorResponse(Exception $ex): void {
+		$this->server->httpResponse->setBody($this->generateBody($ex));
+		$this->server->httpResponse->setStatus(415);
```

Comment:
IIRC, we had to use this response code for client compatibility reasons, right?

Reply:
Now that we have build our own error page and send it by ourselves via the Sabre-Server we could send any response code. But @lennartdohmann and I went again through all the http-status codes and found none that matches better than the 415. Maybe you could argue using 400 but all other codes are of a different topic or have the requested file on the server as subject not the uploaded one.